### PR TITLE
Fix snapshots within a nested crate

### LINKED ISSUE #396: Duplicated snapshots and os error 2 when running a workspace + root crate
### What happened?

When running insta, I get duplicated snapshot reviews, as well as a crash at the end of the snapshot reviewing. In the original (non-minimal) case, I wasn't able to review some of my tests.

I should have reviewed two snapshots, but only was able to review one. The error was the dreaded "No such file or directory (os error 2)".

### Reproduction steps

1. Create a new directory (**not** using `cargo new --lib`)
2. Create a `Cargo.toml`
3. Put this in your `Cargo.toml`:
```toml
[package]
name = "testing2"
version = "0.1.0"

[[test]]
name = "integration"
path = "tests/tests.rs"

[dev-dependencies]
insta = { workspace = true }

[workspace]
members = ["testing"]

[workspace.dependencies]
insta = "1.31"
```
4. Run `cargo new testing --lib` in the project root
5. Put this in `testing/src/lib.rs`
```rust
#[test]
fn it_works() {
    insta::assert_snapshot!("stuff...");
}
```
6. Create the `tests/` directory in the project root
7. Create `tests/tests.rs`
8. Put this in `tests/tests.rs`
```rust
#[test]
fn works() {
    insta::assert_snapshot!("hr");
}
```
9. Run `cargo insta test --review --all`

### Insta Version

1.31.0

### rustc Version

1.71.0

### What did you expect?

I expected to review two tests, and have no crash.

### PR DESCRIPTION
**Pull request by max-sixty**
Fixes #396

### CODE REVIEW

**Comment by max-sixty** [comment]
I've now pushed a slightly broader change, because I was concerned that not passing any packages would fail to include interior packages.

In the existing code, we placed all packages in `LocationInfo`, and then sometimes passed selected packages to functions when required. With the change to fix #396 in the initial commit, we don't pick up all code by running the root package, and without better test coverage I'm not sure whether that then missed parts of the repo.

With the subsequent commits, we now place all packages _we want to run over_ in `LocationInfo`. Making this explicit in a single place means we're less likely to make a mistake.

There's a bit more cloning, but of tiny objects outside any hot loops, and I think worth waiting for #518 etc before optimizing those.

**Comment by regexident** [comment]
@max-sixty It looks like this introduces a regression, which I unfortunately haven't yet been able to reproduce with the [MCVE](https://github.com/mitsuhiko/insta/issues/396#issuecomment-2238483919):

The regression is that while running `insta test` with `--accept --workspace` now works as expected it fails to accept the snapshots when running with a specific package via `--accept --package "package-name"`.

This is the console output when running with `--accept --workspace`:

```terminal
cargo insta test --release --accept --workspace
```

```plain
...

running 3 tests
stored new snapshot <SNIP>/crates/<SNIP>/tests/snapshots/<SNIP>.snap.new
stored new snapshot <SNIP>/crates/<SNIP>/tests/snapshots/<SNIP>.snap.new
...
test snapshots ... ok

...

insta review finished
accepted:
  crates/<SNIP>/tests/snapshots.rs (<SNIP>.snap)
  crates/<SNIP>/tests/snapshots.rs (<SNIP>.snap)
  ...
```

And this is with `--accept --package "<SNIP>"`:

```terminal
cargo insta test --release --accept --package "<SNIP>"
```

```plain
...

running 3 tests
stored new snapshot <SNIP>/crates/<SNIP>/tests/snapshots/<SNIP>.snap.new
stored new snapshot <SNIP>/crates/<SNIP>/tests/snapshots/<SNIP>.snap.new
...
test snapshots ... ok

...
```

Running the latter command (i.e. `--accept --package`) with the currently released version of `cargo-insta` I get the correct behavior.

Looks like it now under-counts in certain circumstances, instead of [double-counting](https://github.com/mitsuhiko/insta/issues/388). 😅

**Comment by max-sixty** [comment]
Thanks! If there's any way of getting a repo that would be really helpful. I'll try to have a think about what might be going on but I suspect it'll be difficult to manage without a repro.

---

One other oddity I noticed using this branch on my work is that snapshots are getting saved with empty headers. I'll try and make a repro myself (it doesn't seem to be all snapshots...)

**Comment by regexident** [comment]
> @max-sixty It looks like this introduces a regression, which I unfortunately haven't yet been able to reproduce with the [MCVE](https://github.com/mitsuhiko/insta/issues/396#issuecomment-2238483919):

I was wrong, it _can_ be reproduced with the MCVE!

I had commented out the root package. Once I reset the shared repo to commit `103e38c`, installed `cargo-insta` from branch `396` and ran `cargo insta test --accept --package "member-crate"` I got this:

```terminal
cargo insta test --accept --package "member-crate"
```

```plain
    Finished `test` profile [unoptimized + debuginfo] target(s) in 0.03s
     Running unittests src/lib.rs (target/debug/deps/member_crate-7d5d5f9bfb1209aa)

running 1 test
stored new snapshot <SNIP>/insta-bug-repro/crates/member-crate/src/snapshots/member_crate__tests__member.snap.new
test tests::test_member ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.14s

   Doc-tests member_crate

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

done: no snapshots to review
```

**Comment by max-sixty** [comment]
Great! Can repro, looking now...

**Comment by max-sixty** [comment]
Fixed! Let me know if that works for you.

(separately I'm thinking about how we can have reasonable integration tests — atm I don't think we're compounding that well, given someone could break this and we wouldn't immediately find out, which then makes changes more difficult. I very much appreciate your patience and help fixing this one...)

File: cargo-insta/src/cargo.rs [hunk 1]
```diff
@@ -50,9 +50,13 @@ pub(crate) fn get_metadata(
     all: bool,
 ) -> Result<Metadata, Box<dyn Error>> {
     let mut cmd = cargo_metadata::MetadataCommand::new();
+
+    // If a manifest path is provided, set it in the command
     if let Some(manifest_path) = manifest_path {
         cmd.manifest_path(manifest_path);
     }
+
+    // TODO: why do we have this? When do we ever want dependencies here?
```
Comment (max-sixty) [inline cargo-insta/src/cargo.rs:59]: All this code is replaced in #536

**Comment by max-sixty** [comment]
This now has a nice integration test with a new integration test approach, in #537

**Comment by max-sixty** [comment]
I would like to merge #537 first and add a test, as I'm not sure I've got the logic right here (even though it seems to pass our hand-tests).

So removing the review request for the moment

**Comment by mitsuhiko** [comment]
I merged #537.

**Comment by mitsuhiko** [comment]
So looking through this I think we need real integration tests here.  I find it hard to follow both existing code and the changes here with confidence that nothing regressed.

**Comment by max-sixty** [comment]
> So looking through this I think we need real integration tests here. I find it hard to follow both existing code and the changes here with confidence that nothing regressed.

Yes agree (re my request to merge #537 first!)

I ended up just integrating everything into #532, because it made the code simpler and easier to debug. So going to close this in favor of that.
